### ISSUE TITLE
Error when running TimeSeriesForestClassifier() function

### ISSUE THREAD

**Issue by i3evelyn** [alan-turing-institute/sktime#160]
**Describe the bug**
Encounter an error when running the code in the example. Specifically, TimeSeriesForestClassifier() function.
PicklingError: Can't pickle <function Pipeline.predict_proba at 0x7fab65da6158>: it's not the same object as sklearn.pipeline.Pipeline.predict_proba

**To Reproduce**
tsf = TimeSeriesForestClassifier(base_estimator=base_estimator, 
                                 n_estimators=100,
                                 criterion='entropy',
                                 bootstrap=True, 
                                 oob_score=True, 
                                 random_state=1)
tsf.fit(X_train, y_train)
tsf.score(X_test, y_test)

**Expected behavior**
Get a score.

**Additional context**

**Versions**
Linux-3.10.0-693.17.1.el7.x86_64-x86_64-with-centos-7.4.1708-Core
Python 3.6.3 |Intel Corporation| (default, Feb 12 2018, 06:37:09) 
[GCC 4.8.2 20140120 (Red Hat 4.8.2-15)]
NumPy 1.17.2
SciPy 1.3.1
sktime 0.3.0

**Comment by mloning** [comment]
Hi and thanks for the bug report! 

I've just run this and it seems to work: 
```python
from sktime.classifiers.compose import TimeSeriesForestClassifier
from sktime.datasets import load_gunpoint
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sktime.transformers.summarise import RandomIntervalFeatureExtractor
from sktime.pipeline import Pipeline
from sktime.utils.time_series import time_series_slope

X_train, y_train = load_gunpoint("TRAIN", return_X_y=True)
X_test, y_test = load_gunpoint("TEST", return_X_y=True)

features = [np.mean, np.std, time_series_slope]
steps = [
    ('transform', RandomIntervalFeatureExtractor(n_intervals='sqrt', features=features)),
    ('clf', DecisionTreeClassifier())
]
base_estimator = Pipeline(steps)

tsf = TimeSeriesForestClassifier(
    base_estimator=base_estimator,
    n_estimators=100,
    criterion='entropy',
    bootstrap=True,
    oob_score=True,
    random_state=1
)
tsf.fit(X_train, y_train)
tsf.score(X_test, y_test)
```

I've seen this error before but thought we've solved it. Could you provide more details on the `base_estimator` and data that you're using?

**Comment by i3evelyn** [comment]
Hi Markus,
Thank you for your replay.
I try to re-run the exact code your show in the Email but it reports the same error I met before.
Information about the environment:

Linux-3.10.0-693.17.1.el7.x86_64-x86_64-with-centos-7.4.1708-Core

Python 3.6.3 |Intel Corporation| (default, Feb 12 2018, 06:37:09)

[GCC 4.8.2 20140120 (Red Hat 4.8.2-15)]

NumPy 1.17.2

SciPy 1.3.1

sktime 0.3.0

sklearn 0.21.3


The reported error content:

PicklingErrorTraceback (most recent call last)

<ipython-input-2-f5d357e3d880> in <module>()

     15 )

     16 tsf.fit(X_train, y_train)

---> 17 tsf.score(X_test, y_test)



/opt/intel/intelpython3/lib/python3.6/site-packages/sklearn/base.py in score(self, X, y, sample_weight)

    355         """

    356         from .metrics import accuracy_score

--> 357         return accuracy_score(y, self.predict(X), sample_weight=sample_weight)

    358

    359



/opt/intel/intelpython3/lib/python3.6/site-packages/sklearn/ensemble/forest.py in predict(self, X)

    543             The predicted classes.

    544         """

--> 545         proba = self.predict_proba(X)

    546

    547         if self.n_outputs_ == 1:



/opt/intel/intelpython3/lib/python3.6/site-packages/sktime/classifiers/compose/ensemble.py in predict_proba(self, X)

    387         n_jobs, _, _ = _partition_estimators(self.n_estimators, self.n_jobs)

    388

--> 389         all_proba = Parallel(n_jobs=n_jobs, verbose=self.verbose)(delayed(e.predict_proba)(X) for e in self.estimators_)

    390

    391         return np.sum(all_proba, axis=0) / len(self.estimators_)



/opt/intel/intelpython3/lib/python3.6/site-packages/joblib/parallel.py in __call__(self, iterable)

    777             # was dispatched. In particular this covers the edge

    778             # case of Parallel used with an exhausted iterator.

--> 779             while self.dispatch_one_batch(iterator):

    780                 self._iterating = True

    781             else:



/opt/intel/intelpython3/lib/python3.6/site-packages/joblib/parallel.py in dispatch_one_batch(self, iterator)

    618

    619         with self._lock:

--> 620             tasks = BatchedCalls(itertools.islice(iterator, batch_size))

    621             if len(tasks) == 0:

    622                 # No more tasks available in the iterator: tell caller to stop.



/opt/intel/intelpython3/lib/python3.6/site-packages/joblib/parallel.py in __init__(self, iterator_slice)

    125

    126     def __init__(self, iterator_slice):

--> 127         self.items = list(iterator_slice)

    128         self._size = len(self.items)

    129



/opt/intel/intelpython3/lib/python3.6/site-packages/sktime/classifiers/compose/ensemble.py in <genexpr>(.0)

    387         n_jobs, _, _ = _partition_estimators(self.n_estimators, self.n_jobs)

    388

--> 389         all_proba = Parallel(n_jobs=n_jobs, verbose=self.verbose)(delayed(e.predict_proba)(X) for e in self.estimators_)

    390

    391         return np.sum(all_proba, axis=0) / len(self.estimators_)



/opt/intel/intelpython3/lib/python3.6/site-packages/joblib/parallel.py in delayed(function, check_pickle)

    181     # using with multiprocessing:

    182     if check_pickle:

--> 183         pickle.dumps(function)

    184

    185     def delayed_function(*args, **kwargs):



PicklingError: Can't pickle <function Pipeline.predict_proba at 0x7f54a0802d08>: it's not the same object as sklearn.pipeline.Pipeline.predict_proba


Thank you!

Regards,
Evelyn


发送自 Windows 10 版邮件<https://go.microsoft.com/fwlink/?LinkId=550986>应用

________________________________

**Comment by mloning** [comment]
I still can't reproduce the error and I'm not sure what causes it. I will later try to run it on a different linux system than we use for unit testing. 

You could try to install the latest development version (cloning and installing from the dev branch as described in the README file) and see if the error still comes up.

**Comment by simone-pignotti** [comment]
Hi, I am experiencing another issue with this function on macOS 10.14.6. It looks easily fixable, attached you can find the log from the code @mloning suggested to run (same behavior for the univariate tutorial) and the package versions in my environment. Thank you for this extremely useful package!
[sktime_bug.log](https://github.com/alan-turing-institute/sktime/files/3926922/sktime_bug.log)
[sktime_env.txt](https://github.com/alan-turing-institute/sktime/files/3926923/sktime_env.txt)

**Comment by mloning** [comment]
Hi @simone-pignotti, thanks for giving us feedback! The code still runs with no errors for me, have you tried it with Python 3.7? There's also a newer sktime version available via pip (0.3.1).

**Comment by simone-pignotti** [comment]
Thank you for the quick answer! Unfortunately I run into the same issue with the following environment (still macOS 10.14.6 and Python 3.7.3, which btw is installed in a conda environment if that matters).
[sktime_env_37.txt](https://github.com/alan-turing-institute/sktime/files/3927315/sktime_env_37.txt)
I'll test on linux asap!

**Comment by mloning** [comment]
Why is there no Python in the environment files that you send? What do you see if you run `conda list -n <your_env> | grep python`?

**Comment by simone-pignotti** [comment]
That was the output of pip freeze, here's the full `conda list`.
[sktime_env_conda.txt](https://github.com/alan-turing-institute/sktime/files/3931574/sktime_env_conda.txt)

**Comment by simone-pignotti** [comment]
Also, same error on Debian 9.

**Comment by mloning** [comment]
Okay, strange, will look into it, but of course would also really appreciate a PR if you have time since I can't really reproduce the error at the moment. :-)

**Comment by simone-pignotti** [comment]
Done, see #206 ! I guess I should have opened a separate issue, since I don't think this fixes @i3evelyn 's original error.
BTW I realized that the problem was the version of scikit-learn. Downgrading it from 0.22.0 to 0.21.3, I was able to get the current release working. On the other end, the pull request needs `scikit-learn>=0.22`, which is why CI currently fails. I hope it's still useful for some future release, while maybe it would be a good idea to change the requirements for the current release to `scikit-learn<0.22`.
I take the chance to raise another requirements-related issue: after creating a conda environment, `pip install sktime` will fail on macOS because `numpy` and `Cython` are not available and pip won't install them automatically. This doesn't happen on linux.
Thank you again, let me know if you prefer that I open a separate issue for that.
